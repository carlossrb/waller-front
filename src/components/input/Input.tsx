import React, { FC } from 'react';
import cn from 'classnames';

export const sizes = {
  xs: 'text-font300 px-200 py-100',
  sm: 'text-font300 px-200 py-200',
  md: 'text-font300 px-200 py-300',
  lg: 'text-font300 px-200 py-400',
};

export type InputProps = {
  size?: keyof typeof sizes;
  error?: boolean;
  autoFocus?: boolean;
  onChange?: (v: string) => void;
  value?: string;
  left?: React.ReactNode;
} & Omit<JSX.IntrinsicElements['input'], 'size' | 'onChange'>;

export const Input: FC<InputProps> = ({
  size = 'md',
  error = false,
  className,
  onChange,
  value,
  autoFocus = false,
  left,
  ...props
}) => {
  const rootClasses = cn(
    'flex w-full transition-all rounded-200 ring-2 ring-gray400',
    'focus-within:ring-gray200',
    {
      'bg-backgroundColor100 text-gray200': !error,
      'ring-danger400 focus-within:ring-danger200': error,
    },
    className
  );
  const inputClasses = cn(
    'w-full rounded-200 placeholder-gray200 block bg-backgroundColor100 focus:outline-none focus:bg-backgroundColor',
    sizes[size]
  );
  const leftClasses = cn('flex space-x-400 text-gray200', sizes[size]);

  return (
    <div className={rootClasses}>
      {left && (
        <div data-testid="input-left-content" className={leftClasses}>
          {left}
        </div>
      )}
      <input
        aria-label="input-component"
        autoFocus={autoFocus}
        {...props}
        value={value}
        onChange={(e) => onChange?.(e.target.value)}
        className={inputClasses}
      />
    </div>
  );
};
