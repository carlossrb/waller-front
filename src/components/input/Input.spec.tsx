import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Input } from './Input';

describe('Input', () => {
  /**
   * Style tests.
   */

  it('should render vanilla input', () => {
    const { getByLabelText } = render(<Input />);
    expect(getByLabelText('input-component')).toMatchSnapshot();
  });

  it('should render error input', () => {
    const { getByLabelText } = render(<Input error />);
    expect(getByLabelText('input-component')).toMatchSnapshot();
  });

  it('should render xs input', () => {
    const { getByLabelText } = render(<Input size="xs" />);
    expect(getByLabelText('input-component')).toMatchSnapshot();
  });

  it('should render sm input', () => {
    const { getByLabelText } = render(<Input size="sm" />);
    expect(getByLabelText('input-component')).toMatchSnapshot();
  });

  it('should render lg input', () => {
    const { getByLabelText } = render(<Input size="lg" />);
    expect(getByLabelText('input-component')).toMatchSnapshot();
  });

  it('should render input with focus', () => {
    const { getByLabelText } = render(<Input autoFocus />);
    expect(getByLabelText('input-component')).toMatchSnapshot();
  });

  /**
   * Use case / behavior tests.
   */

  it('should a R$ in front of the input', () => {
    const { getByTestId } = render(<Input left={'R$'} />);
    expect(getByTestId('input-left-content')).toHaveTextContent('R$');
  });

  test('should modify value with waller', () => {
    const onChange = jest.fn();
    const { getByLabelText } = render(<Input onChange={onChange} />);
    const input = getByLabelText('input-component');
    fireEvent.change(getByLabelText('input-component'), { target: { value: 'waller' } });
    expect((input as any).value).toBe('waller');
  });
});
