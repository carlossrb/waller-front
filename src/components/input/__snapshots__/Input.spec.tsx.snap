// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render error input 1`] = `
<input
  aria-label="input-component"
  class="w-full rounded-200 placeholder-gray200 block bg-backgroundColor100 focus:outline-none focus:bg-backgroundColor text-font300 px-200 py-300"
  value=""
/>
`;

exports[`Input should render input with focus 1`] = `
<input
  aria-label="input-component"
  class="w-full rounded-200 placeholder-gray200 block bg-backgroundColor100 focus:outline-none focus:bg-backgroundColor text-font300 px-200 py-300"
  value=""
/>
`;

exports[`Input should render lg input 1`] = `
<input
  aria-label="input-component"
  class="w-full rounded-200 placeholder-gray200 block bg-backgroundColor100 focus:outline-none focus:bg-backgroundColor text-font300 px-200 py-400"
  value=""
/>
`;

exports[`Input should render sm input 1`] = `
<input
  aria-label="input-component"
  class="w-full rounded-200 placeholder-gray200 block bg-backgroundColor100 focus:outline-none focus:bg-backgroundColor text-font300 px-200 py-200"
  value=""
/>
`;

exports[`Input should render vanilla input 1`] = `
<input
  aria-label="input-component"
  class="w-full rounded-200 placeholder-gray200 block bg-backgroundColor100 focus:outline-none focus:bg-backgroundColor text-font300 px-200 py-300"
  value=""
/>
`;

exports[`Input should render xs input 1`] = `
<input
  aria-label="input-component"
  class="w-full rounded-200 placeholder-gray200 block bg-backgroundColor100 focus:outline-none focus:bg-backgroundColor text-font300 px-200 py-100"
  value=""
/>
`;
